app/views/tasks/show.html.slim

```diff
@@ -573,10 +573,9 @@
             i.fa-regular.fa-clone
             =< t('common.button.duplicate')
 
-      - if policy(@task).generate_test?
```

Comment:
With the policy check, the button only appears if you are a user and with an API key. but since I am handling now the missing or invalid key, I want the feature to be always visible. does that make sense?

Reply:
In that case I would hide the button, if the user is not able to `update` the task and disable it with an informative tooltip if they have no key.

Reply:
I am a bit confused as to why we need to do this?
Since I am handling the fact that a user might have an empty key, why do we need to disable it? There are alerts for missing or invalid keys already in place.

Also by adding the extra condition in the policy, my alert is no longer visible because I then get 'you are not authorized for this' alert

I feel like my approach is handling all cases better. but maybe I am misunderstanding something?

P.S I have tackled all your other concerns and will push the changes now. :)

Reply:
I think my approach would be more intuitive, since the user can immediately see what he is able to do. Showing him the "generate test" button when is he not even able to edit the task, doesn't make sense to me. And in the same way suggesting, that you can use the feature (by showing the button) without having the ability (since the API-Key is missing) does not feel right either. 
It might make sense to show that the feature exists, that's why I suggested disabling (greying out) the button and adding a tooltip.

Reply:
I feel like this is more about how each of us would like to see the feature look like at this point :D 
I think both solutions are viable and which one is more intuitive would have to be decided by the user (future experiment)

I am okay with disabling the button with some kind of message that a key is needed to use it. Could you help show how this can be done? i have no idea how to make a tooltip even with the help of chatgpt it doesn't make sense :D

Reply:
Additionally I was planning to add a block for the generate unit test instead of the button to make it more visible with a clear message that this is a beta-feature. sebastian suggested something like that in the first PR.

Reply:
for the disabling you could look at https://github.com/openHPI/codeharbor/blob/d41c1ffd09f3900e438b60d2f0f5ce2e4b9b0a22/app/views/groups/show.html.slim#L88 and the tooltip: https://github.com/openHPI/codeharbor/blob/3030671711248276c7c46d8df8730d23c230252b/app/views/tasks/_tasks.html.slim#L71

Reply:
> Additionally I was planning to add a block for the generate unit test instead of the button to make it more visible with a clear message that this is a beta-feature. sebastian suggested something like that in the first PR.

I'm not sure what you mean by "using a block"

Reply:
a block like the `add to collection` one

Reply:
Ah I see you discussed it here right? https://github.com/openHPI/codeharbor/pull/1261#discussion_r1656853037
I would add it to the list in #1492 and tackle it separately

Reply:
Yes exactly. It will be done separately from this PR

Reply:
@kkoehn is this how the code should look like:

```
    = link_to generate_test_task_path(@task), method: :post, class: "btn btn-important #{'disabled' unless current_user.openai_api_key}", data: {bs_toggle: 'tooltip', bs_placement: 'top', bs_title: t('.button.api_key_required')} do
      i.fa-solid.fa-wand-magic-sparkles
      = t('.button.generate_test')
```

The button is still clickable though like that, what is missing?

Reply:
when I do like this:

```
      - if policy(@task).generate_test?
        = link_to generate_test_task_path(@task), method: :post, class: 'btn btn-important' do
          i.fa-solid.fa-wand-magic-sparkles
          = t('.button.generate_test')
      - else
        = link_to '#', method: :post, class: 'btn btn-important disabled', data: {bs_toggle: 'tooltip', bs_placement: 'top', bs_title: t('.button.api_key_required')} do
          i.fa-solid.fa-wand-magic-sparkles
          = t('.button.generate_test')
```

The button is disabled when the key is not provided, but no tooltip appears

Reply:
The issue seems to arise because disabled elements often don’t trigger hover or focus events, which are required to show tooltips.

In my opinion, just disabling the button and not showing why is not ideal at all and I wouldn't go with this approach if that's the case

Reply:
I figured it out by not linking the button to anything without disabling it: is that a viable solution:

```
      - if policy(@task).generate_test?
        = link_to generate_test_task_path(@task), method: :post, class: 'btn btn-important' do
          i.fa-solid.fa-wand-magic-sparkles
          = t('.button.generate_test')
      - else
        = link_to '#', class: 'btn btn-important', data: {bs_toggle: 'tooltip', bs_placement: 'top', bs_title: t('.button.api_key_required')} do
          i.fa-solid.fa-wand-magic-sparkles
          = t('.button.generate_test')
```

Pushing all changes now :)

Reply:
I don't want to jump into this discussion here, but the tooltips @kkoehn is referring to are working for a disabled button. [Check this task](https://codeharbor.openhpi.de/tasks/202?locale=en) in a private browsing window and you'll see:

<img width="245" alt="Bildschirmfoto 2024-07-08 um 23 49 29" src="https://github.com/openHPI/codeharbor/assets/7300329/e2fc3e86-b48b-4856-8eea-eabb432f2935">

Reply:
hmm that's weird then, it's not working for me when I test it locally. could you share the code for this please? :)

Reply:
I did this and it didn't work:

```
      - if policy(@task).generate_test?
        = link_to generate_test_task_path(@task), method: :post, class: 'btn btn-important' do
          i.fa-solid.fa-wand-magic-sparkles
          = t('.button.generate_test')
      - else
        = link_to '#', class: 'btn btn-important disabled', data: {bs_toggle: 'tooltip', bs_placement: 'top', bs_title: t('.button.api_key_required')} do
          i.fa-solid.fa-wand-magic-sparkles
          = t('.button.generate_test')
```

Only when I remove the `disabled` part, it works

Reply:
Are you referring to this one? 

https://github.com/openHPI/codeharbor/blob/3030671711248276c7c46d8df8730d23c230252b/app/views/tasks/show.html.slim#L566-L574

Reply:
Yes. sorry I didn't see that in the file :) 
I was able to make it work in a similar way. thanks for your help

Reply:
Your solution was almost what I wanted. I made a commit to which implements it all the way - please check, if you find it acceptable